### PR TITLE
Persisted favorite parameter types and major deprecation filtering refactor

### PR DESCRIPTION
**Pull request by alexatrhea**
### Prerequisites

- [ ] I have written a descriptive pull-request title
- [ ] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [ ] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [ ] I have provided test coverage for my change (where applicable)

### Description

- Fixed bug where scrollbar dragging would become non-functional all of a sudden.
- RDL ribbon and browser cleanup, styles, ribbon groups spelling etc.
- Left and Right regions are now TabbedGroups so that browsers do not pile on indefinitely blocking the screen.
- Element Definition browser is now in central group by default.
- Added Favoriting functionality to the ParameterTypes browser. Use context menu to toggle whether a parameter type is favorited. Use button in the toolbar to toggle filter.
- Complete refactor of the filtering system to make deprecatable and favoritable filters work correctly together.
- Overhaul of browsers that make use of deprecated things to make it work with filtering (both grids and trees).

### CODE REVIEW

File: BasicRdl/Views/CategoryBrowser.xaml.cs [hunk 1]
```diff
@@ -47,8 +47,13 @@ public CategoryBrowser(bool initializeComponent)
             if (initializeComponent)
             {
                 this.InitializeComponent();
-                FilterStringService.FilterString.AddGridControl(this.CategoriesGridControl);
+                this.Control = this.CategoriesGridControl;
```
Comment (lxatrhea) [inline BasicRdl/Views/CategoryBrowser.xaml.cs]: I think the name "Control" is too generic.
When you look at the IPanelFilterableDataGridView interface, it's a logical thing, but in the implementations I think 
FilterableControl (or something like that) says more.

File: BasicRdl/Views/ParameterTypesBrowser.xaml [hunk 2]
```diff
@@ -31,7 +31,7 @@
             <RowDefinition Height="*" />
         </Grid.RowDefinitions>
 
-        <views:CommonThingControl GridView="{Binding ElementName=View}" />
+        <views:CommonThingControl GridView="{Binding ElementName=View}" IsFavoriteToggleEnabled="True" />
```
Comment (lxatrhea) [inline BasicRdl/Views/ParameterTypesBrowser.xaml]: Enabled, or visible?

File: BasicRdl/Views/Ribbons/CategoryRibbon.xaml [hunk 3]
```diff
@@ -1,33 +1,31 @@
...
 
-        <dxb:BarSplitButtonItem.PopupControl>
-            <dxb:PopupMenu ItemLinksSource="{Binding OpenSessions}" />
-        </dxb:BarSplitButtonItem.PopupControl>
-    </dxb:BarSplitButtonItem>
-</dxr:RibbonPageGroup>
+    <dxb:BarSplitButtonItem.PopupControl>
+        <dxb:PopupMenu ItemLinksSource="{Binding OpenSessions}" />
+    </dxb:BarSplitButtonItem.PopupControl>
+</dxb:BarSplitButtonItem>
```
Comment (lxatrhea) [inline BasicRdl/Views/Ribbons/CategoryRibbon.xaml]: New line

File: BasicRdl/Views/Ribbons/BasicRdlRibbon.xaml [hunk 4]
```diff
@@ -5,22 +5,29 @@
...
+    </dxr:RibbonPageGroup>
+    <dxr:RibbonPageGroup Caption="References">
+        <views:UnitPrefixRibbon />
+        <views:ConstantBrowserRibbon />
+        <views:GlossaryBrowserRibbon />
+        <views:FileTypeBrowserRibbon />
+        <views:ReferenceSourceRibbon />
+    </dxr:RibbonPageGroup>
+
 </ribbon:ExtendedRibbonPage>
```
Comment (lxatrhea) [inline BasicRdl/Views/Ribbons/BasicRdlRibbon.xaml]: new line

File: CDP4Composition/Mvvm/BrowserViewModelBase.cs [hunk 5]
```diff
@@ -869,6 +869,7 @@ public virtual void PopulateContextMenu()
                 this.ContextMenu.Add(categoriesMenu);
             }
 
+            // TODO: This is really only applicable to expandable things, and certainly does not even need to be in grids.
```
Comment (lxatrhea) [inline CDP4Composition/Mvvm/BrowserViewModelBase.cs]: Is there an issue for this?

File: CDP4SiteDirectory/ViewModels/OrganizationBrowser/OrganizationBrowserViewModel.cs [hunk 6]
```diff
@@ -113,7 +113,7 @@ private void UpdatePersonRows(OrganizationRowViewModel organizationRow)
 
             foreach (var addedPerson in addedPersons)
             {
-                var row = new OrganizationBrowser.PersonRowViewModel(addedPerson, this.Session, this);
+                var row = new OrganizationBrowser.PersonRowViewModel(addedPerson, this.Session, organizationRow);
```
Comment (lxatrhea) [inline CDP4SiteDirectory/ViewModels/OrganizationBrowser/OrganizationBrowserViewModel.cs]: I don't understand. Is this correct?

Comment (alexatrhea) [reply CDP4SiteDirectory/ViewModels/OrganizationBrowser/OrganizationBrowserViewModel.cs]: Yes, `this` in this case is the `OrganizationBrowserViewModel` which is incorrect vm for the `ContainerRow` of the `PersonRowViewModel`.

File: CDP4Composition/Mvvm/BrowserViewModelBase.cs (hunk 5 above)
Comment (alexatrhea) [reply CDP4Composition/Mvvm/BrowserViewModelBase.cs]: #254

File: BasicRdl/ViewModels/ConstantsBrowserViewModel.cs [hunk 7]
```diff
@@ -56,26 +55,27 @@ public class ConstantsBrowserViewModel : BrowserViewModelBase<SiteDirectory>, IP
...
-            : base(siteDir, session, thingDialogNavigationService, panelNavigationService, dialogNavigationService, pluginSettingsService)
+        public ConstantsBrowserViewModel(ISession session, SiteDirectory siteDir,
+            IThingDialogNavigationService thingDialogNavigationService, IPanelNavigationService panelNavigationService,
+            IDialogNavigationService dialogNavigationService, IPluginSettingsService pluginSettingsService)
+            : base(siteDir, session, thingDialogNavigationService, panelNavigationService, dialogNavigationService,
+                pluginSettingsService)
         {
             this.Caption = string.Format("{0}, {1}", PanelCaption, this.Thing.Name);
-            this.ToolTip = string.Format("{0}\n{1}\n{2}", this.Thing.Name, this.Thing.IDalUri, this.Session.ActivePerson.Name);
+            this.ToolTip = string.Format("{0}\n{1}\n{2}", this.Thing.Name, this.Thing.IDalUri,
```
[the remark this answers is not in the document]
Comment (alexatrhea) [reply BasicRdl/ViewModels/ConstantsBrowserViewModel.cs]: Ideally it has to be fixed in too many places so I leave this for later. some sort of general cleanup has to take place.

File: BasicRdl/ViewModels/CategoryBrowserViewModel.cs [hunk 8]
```diff
@@ -53,24 +53,25 @@ public class CategoryBrowserViewModel : BrowserViewModelBase<SiteDirectory>, IPa
...
-            : base(thing, session, thingDialogNavigationService, panelNavigationService, dialogNavigationService, pluginSettingsService)
+        public CategoryBrowserViewModel(ISession session, SiteDirectory thing,
+            IThingDialogNavigationService thingDialogNavigationService, IPanelNavigationService panelNavigationService,
+            IDialogNavigationService dialogNavigationService, IPluginSettingsService pluginSettingsService)
+            : base(thing, session, thingDialogNavigationService, panelNavigationService, dialogNavigationService,
+                pluginSettingsService)
         {
             this.Caption = string.Format("{0}, {1}", PanelCaption, this.Thing.Name);
-            this.ToolTip = string.Format("{0}\n{1}\n{2}", this.Thing.Name, this.Thing.IDalUri, this.Session.ActivePerson.Name);
+            this.ToolTip = string.Format("{0}\n{1}\n{2}", this.Thing.Name, this.Thing.IDalUri,
```
[the remark this answers is not in the document]
Comment (alexatrhea) [reply BasicRdl/ViewModels/CategoryBrowserViewModel.cs]: answered in another comment.

File: CDP4Composition/Services/FavoritesService/FavoritesService.cs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            transaction.CreateOrUpdate(preference);
+
+            try
+            {
+                var operationContainer = transaction.FinalizeTransaction();
+                await session.Write(operationContainer);
+            }
+            catch (Exception ex)
+            {
+                LogManager.GetLogger(typeof(FavoritesService).FullName).Error("The inline update operation failed: {0}", ex.Message);
```
Comment (samatrhea) [inline CDP4Composition/Services/FavoritesService/FavoritesService.cs]: logger is instantiated once on top of the class, no need to get it for every exception

**Comment by alexatrhea** [comment]
Closing and merging from command line.

**Pull request closed without merging**